async/cohttp_async.ml

```diff
@@ -156,25 +156,60 @@ end
 
 module Client = struct
 
+  let read_request ic =
+    Response.read ic >>| function
+    | `Eof -> failwith "Connection closed by remote host"
+    | `Invalid reason -> failwith reason
```

Comment:
@rgrinberg Yes that's fine. Would your thought be that another PR would introduce a new error handling mechanism? As you point out the current error handling strategy is not very useful in the wild. It seems that one wants a ` (Response.t * Body.t) Or_error.t Pipe.Reader.t Deferred.t` instead which would allow one to figure out what is going on with each response.